Extension/src/LanguageServer/configurations.ts

```diff
@@ -708,6 +708,26 @@ export class CppProperties {
 
             configuration.browse.limitSymbolsToIncludedHeaders = this.updateConfigurationStringOrBoolean(configuration.browse.limitSymbolsToIncludedHeaders, settings.defaultLimitSymbolsToIncludedHeaders, env);
             configuration.browse.databaseFilename = this.updateConfigurationString(configuration.browse.databaseFilename, settings.defaultDatabaseFilename, env);
+
+            // If there is no c_cpp_properties.json, there are no relevant C_Cpp.default.* settings set,
+            // and there is only 1 registered custom config provider, default to using that provider.
+            const providers: CustomConfigurationProviderCollection = getCustomConfigProviders();
+            if (providers.size === 1
+                && !this.propertiesFile
+                && !settings.defaultCompilerPath
```

Comment:
If the customer sets `C_Cpp.default.compilerPath` to empty string (to disable compiler querying), I think this will do the wrong thing.  And I'm wondering if it's safer to use something like `settingsTracker.getUserModifiedSettings` and check that against the settings names in case we ever change defaults.

Reply:
This seems simple to fix by checking for an empty compilerPath here.  getUserModifiedSettings would also gather settings that don't apply to the configuration.  There are various other checks in this function of these individual settings.  We could move some or all of that into new SettingsTracker functions, but that seems like a can-of-worms, and may not wind up being simpler.

Reply:
I think it's safest to do explicit comparison (`!==` or `===`) with the default settings values for all the settings instead of using `!` then.

Reply:
It's already comparing all string based setting with `===` against `''`, instead of using `!`.  VS Code does not represent string settings using null or undefined, it actually returns an empty string if those settings are not set.  This is related to the settings UI, in which an empty text field and an unset setting are equivalent.  compilerPath is our only current exception (which results in this setting not being editable using a test field in the settings UI, which is unfortunate), and for that I'm now treating `""` differently from unset.  For ALL other strings, it is not using `!`, it's checking against empty string, which is correct.  The remaining uses of `!` are for arrays, and that seems preferable over explicitly checking against both null and undefined.  Unlike strings, when an array is defined but empty, `!` will return false.  So, if there is something that should be changed here to be more explicit, I don't see it.  Let me know.  :)